api/envoy/extensions/filters/http/api_key_auth/v3/api_key_auth.proto

```diff
@@ -28,27 +26,85 @@ option (xds.annotations.v3.file_status).work_in_progress = true;
 // .. code-block:: yaml
 //
 //    authentication_header: "X-API-KEY"
-//    keys:
-//      inline_string: |-
-//        clientID1:apiKey1
-//        clientID2:apiKey2
+//    credentials:
+//      entries:
+//      - key: 09876abcdefg
+//        client_id: user
+//    key_sources:
+//      entries:
+//      - header: "X-API-KEY"
 //
-message APIKeyAuth {
-  // keys used to authenticate the client.
-  // It should be a map of clientID to apiKey.
-  // The clientID serves solely for identification purposes and isn't used for authentication.
-  config.core.v3.DataSource keys = 1 [(udpa.annotations.sensitive) = true];
+message ApiKeyAuth {
+  // The credentials that are used to authenticate the clients.
+  Credentials credentials = 1 [(udpa.annotations.sensitive) = true];
+
+  // The key sources to fetch the key from the coming request.
+  KeySources key_sources = 2;
+}
 
-  // The header name to fetch the key.
-  // If multiple values are present in the given header, the filter rejects the request.
-  // Only one of authentication_header, authentication_query, or authentication_cookie should be set.
-  string authentication_header = 2;
+// API key auth configuration of per route or per virtual host or per route configuration.
+message ApiKeyAuthPerRoute {
+  // Route specific APIKeyAuth configuration. This is optional and could be used to override the
+  // filter level ApiKeyAuth configuration **partly**.
+  //
+  // For example, if route specific ``authentication_header`` is set, then the filter will fetch
+  // the key from route specific ``authentication_header``. But the filter level ``keys`` will
+  // still be used. If route specific non-empty ``keys`` is set, then the filter will use the
+  // route specific ``keys`` to authenticate the client but the filter level ``authentication_header``
+  // will still be used. If both ``authentication_header`` and ``keys`` are set, then the filter
+  // will use the route specific ``authentication_header`` and ``keys`` to authenticate the client.
+  ApiKeyAuth override_config = 1;
 
-  // The query parameter name to fetch the key.
-  // Only one of authentication_header, authentication_query, or authentication_cookie should be set.
-  string authentication_query = 3;
+  // A list of client ids that are allowed to access the routes of the scope. If the list is empty,
+  // then all authenticated clients are allowed. This provides very limited but simple authorization.
+  // If more complex authorization is required, then use the :ref:`HTTP RBAC filter
+  // <config_http_filters_rbac>` instead.
+  repeated string allowed_clients = 2;
```

Comment:
I find this field to be somewhat confusing, as there isn't a list of allowed clients in the non-route-override case.

Reply:
There are two fields in the `ApiKeyAuthPerRoute`. One `override_config` field which is used to selectively override the filter level config. One `allowed_clients` field which is route-specific config (only in the `ApiKeyAuthPerRoute`) that only could be configured on the route/vhost. The `allowed_clients` is used to limit the accessing to specific route/vhost.

Reply:
They are independent won't effect each others. The `override_config` is used to override the filter level configuration. And the `allowed_clients` is used to control accessing to the route.

The filter will try to authenticate the client by the `override_config` (if exists) and filter level configuration first. After the client is authenticated, we will try to check if it's limited by the `allowed_clients` (if exists and non-empty).

Reply:
> The filter will try to authenticate the client by the `override_config` (if exists) and filter level configuration first. After the client is authenticated, we will try to check if it's limited by the `allowed_clients` (if exists and non-empty).

IMHO this is somewhat fragile, and may lead to inconsistencies when new fields are added.
Specifically, IIUC, if both fields are set, the client needs both in the credentials list inside the `override_config` and in the `allowed_clients` list.
I understand the motivation behind adding one of the fields, but I'm not sure I fully grok when would adding both fields be useful.
To be more concrete, looking at the [example below](https://github.com/envoyproxy/envoy/pull/36968/files#diff-ce204a200ca6e35d0514de28dd809c21873a818bd21da9d529d2435c2de654efR28-R39), it is unclear to me why the {one_key, one_client} pair is useful in the override config (other than maybe receiving a different HTTP response code).

Reply:
I konw your worry. Because I initially share complete same thinking with you. This is why the first vesion of the API (I reviewed that.) has only one message `ApiKeyAuth`. At that time, I think the route and filter could shared complete same configuration structure.

But now, I will think the usual composition is the `filter level credentials` and route level `allowed_clients`. And for multiple talent gateway (or some special scenarios where need route specific key-sources), the route level `override_config` will helps.

The example may not be good enough. I will update it.

Reply:
> I initially think we could re-use the ApiKeyAuth as the per route/vhost filter configuration. But after a thinking from control planes' perspective, I change my mind.
> Considering that the api keys may be changed or updated for a client id/consumer id, it would be better to share same credential list for all route. So, when credential updates occur, the control plane needn't to iterate all routes to find the routes that need to update.

Biscally, the gateway product will never let the users to configure their credencials like a normal filter configuration.  The crendentials will be managed by an independent module and the users will update the crendentials at that module. Then that module will trigger a update to the filter configuration.

When we need limit the clients that could access specific routes, the users will specific some clients explicity.

But if we use the override crendentials at the routes to do that, then when the user updated their crendentials, the control plane need to iterate all routes and compute which routes need to be updated.

Using the central credential list and a route level simple `allowed_clients` would make the processing be much simpler.

Reply:
I understand why `allowed_clients` by itself (without `override_config`) is useful, and I think that addresses the use-case you described here.

What I still fail to understand is when both `override_config` and `allowed_clients` could be used.

Reply:
> What I still fail to understand is when both override_config and allowed_clients could be used.

I also think this rarely happened.

Reply:
Got it.